File: java/filterservice/src/main/java/client/GeofenceClient.java
package client;

import akka.actor.ActorSystem;
import akka.dispatch.ExecutionContexts;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.unmarshalling.Unmarshaller;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;

import java.util.concurrent.CompletionStage;

public class GeofenceClient {

    private String geofenceUri;
    private ActorSystem system;
    private Materializer materializer;

    public GeofenceClient(String geofenceUri, ActorSystem system) {
        this.geofenceUri = geofenceUri;
        this.system = system;
        this.materializer = ActorMaterializer.create(system);
    }

    public CompletionStage<Boolean> isInGeofence(Double lat, Double lon) {
        return Http.get(system)
            .singleRequest(HttpRequest.create(geofenceUri + "?lat=" + lat + "&lon=" + lon))
            .thenCompose(response -> Unmarshaller.entityToString().unmarshal(response.entity(), ExecutionContexts.global(), materializer))
            .thenApply(Boolean::parseBoolean);
    }
}

File: filter-service/src/main/scala/client/GeofenceClient.scala
package client

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.Uri.Query
import akka.http.scaladsl.model._
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.Materializer

import scala.concurrent.Future

case class GeofenceClient(geofenceUri: Uri)(implicit val system: ActorSystem, materializer: Materializer) {

  import system.dispatcher

  def isInGeofence(lat: Double, lon: Double): Future[Boolean] = {
    val getUri = geofenceUri
      .withQuery(Query(
          "lat" -> lat.toString,
          "lon" -> lon.toString
        )
      )

    import akka.http.scaladsl.unmarshalling.Unmarshaller._

    for {
      httpResponse: HttpResponse <- Http().singleRequest(HttpRequest(
        uri = getUri,
        method = HttpMethods.GET
      ))
      string <- Unmarshal(httpResponse).to[String]
      response <- Unmarshal(string).to[Boolean]
    } yield response
  }
}
